src/lsm/manifest_level.zig

```diff
@@ -374,20 +423,216 @@ pub fn ManifestLevelType(
         }
 
         /// The function is only used for verification; it is not performance-critical.
-        pub fn contains(level: Self, table: *const TableInfo) bool {
+        /// The table provided could either be a pointer to the table (from_level) in the
+        /// ManifestLevel or a pointer to a copy of that table (copy).
+        pub fn contains(level: Self, table_union: TableInfoPtrConst) bool {
             assert(constants.verify);
-
+            const table = switch (table_union) {
+                .copy => table_union.copy,
+                .from_level => table_union.from_level,
+            };
             var level_tables = level.iterator(.visible, &.{
                 table.snapshot_min,
             }, .ascending, KeyRange{
                 .key_min = table.key_min,
                 .key_max = table.key_max,
             });
             while (level_tables.next()) |level_table| {
-                if (level_table.equal(table)) return true;
+                switch (table_union) {
+                    // If the table provided is a pointer to a copy of the table,
+                    // we check for equality of all important table properties.
+                    .copy => {
+                        if (level_table.equal(table)) return true;
+                    },
+                    // If the table provided is a pointer to the table in the ManifestLevel,
+                    // we check for equality of the pointers. This is to ensure
+                    // that the provided table is a valid pointer in the ManifestLevel.
+                    .from_level => {
+                        if (level_table == table) return true;
+                    },
+                }
             }
             return false;
```

Comment:
Ah, maybe I misunderstood what you said in https://github.com/tigerbeetledb/tigerbeetle/pull/875#discussion_r1228536880? I thought we wanted to validate whether a pointer that the invoker of `update_table` claims is `from_level` is really a *TableInfo within ManifestLevel. Wouldn't that be useful to check?

Or maybe I didn't completely understand what you meant in this commit! Could you please explain it a bit more? :D

Reply:
> I thought we wanted to validate whether a pointer that the invoker of update_table claims is from_level is really a *TableInfo within ManifestLevel. Wouldn't that be useful to check?

Yes, that's right.

To expand on:

> Can we assert(table_union == .copy); here? i.e. if it is from_level, then we would expect to have found it in our level.


We are about to `return false`, i.e. "the manifest level does not contain the table that was passed in".
But if `table_union == .from_level` (i.e. an internal pointer), then by definition the pointer should be in our level.

Reply:
> We are about to return false, i.e. "the manifest level does not contain the table that was passed in".
But if table_union == .from_level (i.e. an internal pointer), then by definition the pointer should be in our level.

Ah, understood! I was under the impression that returning `False` here is valid even if `table_union == .from_level`, because maybe the TableInfo* pointer doesn't exist in the ManifestLevel due to rearrangements in the Segmented Array causing the pointer to get invalidated?

Reply:
> I was under the impression that returning False here is valid even if table_union == .from_level, because maybe the TableInfo* pointer doesn't exist in the ManifestLevel due to rearrangements in the Segmented Array causing the pointer to get invalidated?

Can you elaborate on this case? If the segmentedarray was rearranged then it the caller shouldn't be using `from_level`, right?

Reply:
Yes, the caller shouldn't be using ManifestLevel references if the Segmented Array was rearranged! I thought that returning `False` from `ManifestLevel.contains` would be a good way to catch issues where the caller is using`.from_level` but the pointer being used has been invalidated...